### PR TITLE
Expose timeout query param for search requests

### PR DESCRIPTION
**Pull request by coszio**
Continues upon #2293 and solves #2623

Exposes timeout param for search requests.
- in query params for REST
- in top level args for gRPC
 
Affects at local shard level and internal client request. For group requests, it affects at the `GroupBy` level

### Checklist:
- [x] Expose for search and batch search
- [x] Expose for recommend and batch recommend
- [x] Expose for group (search/recommend) requests
- [x] Implement timeout for `GroupBy`
- [x] Make tests

## Caveats

Search and recommend requests return a `500 Service Error` status because of [this conversion](https://github.com/qdrant/qdrant/blob/e7d53aac8e907339eb94a2e61c1269c99136671d/lib/collection/src/shards/replica_set.rs#L993), even when it is just a timeout error. This is different from what grouping requests return now, which is a standard `408 Request Timeout`. We can make both return `500` if desired, but I feel we should fix the former case.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by timvisee — COMMENTED** [review]
Nice work! I haven't been able to actually test it jet but did want to post my review comments thus far.

I'll try to test tomorrow.

**Comment by timvisee** [comment]
Sadly, the recent <https://github.com/qdrant/qdrant/pull/2761> and <https://github.com/qdrant/qdrant/pull/2762> merges brought some conflicts.

File: src/actix/api/read_params.rs [hunk 1]
```diff
@@ -8,6 +10,15 @@ pub struct ReadParams {
     #[serde(default, deserialize_with = "deserialize_read_consistency")]
     #[validate]
     pub consistency: Option<ReadConsistency>,
+    /// If set, overrides global timeout for this request. Unit is seconds.
+    #[validate(range(min = 1))]
+    pub timeout: Option<u64>,
```
[the remark this answers is not in the document]
Comment (coszio) [reply src/actix/api/read_params.rs:15]: Makes little difference, but I like using it, changed

File: openapi/openapi-main.ytt.yaml [hunk 2]
```diff
@@ -295,6 +295,12 @@ paths:
           required: false
           schema:
             $ref: "#/components/schemas/ReadConsistency"
+        - name: timeout
+          in: query
+          description: If set, overrides global timeout for this request. Unit is seconds.
+          required: false
+          schema:
+            type: integer
```
[the remark this answers is not in the document]
Comment (coszio) [reply openapi/openapi-main.ytt.yaml:303]: True! thanks for adding this

**Comment by coszio** [comment]
An update:
@generall noted that the current changes do not extend the default timeout for remote shard communication. 

Here's what I've tried:
- The way it works with these changes is that we add a timeout to the `Request` the internal grpc client makes to the remote shard.
- We also have available our method of [`with_channel_timeout`](https://github.com/qdrant/qdrant/blob/72c10d2ffe49d8a355b711c5affd115d199f9c9a/lib/api/src/grpc/transport_channel_pool.rs#L255), but after using it, it doesn't solve our problem.

In first scenario we are limited by the default max_timeout of `with_channel_timeout` so we sometimes get an error like
```rust
"Timeout {}ms reached for uri: {}"

or

"Tonic status error: The operation was cancelled"
```

In second scenario, we only get the second error.

So what's happening? `with_channel_timeout` is only [applying a separate timeout](https://github.com/qdrant/qdrant/blob/72c10d2ffe49d8a355b711c5affd115d199f9c9a/lib/api/src/grpc/transport_channel_pool.rs#L235C13-L237C13) from the [actual channel timeout](https://github.com/qdrant/qdrant/blob/72c10d2ffe49d8a355b711c5affd115d199f9c9a/lib/api/src/grpc/transport_channel_pool.rs#L106).

I propose to deal with that complexity in another PR, and we leave the constraint that, for remote shards, timeout will be the lowest of global or query param.

My second proposal is to refactor to get rid of the custom layer of timeout of `with_channel_timeout` so that the only way to affect it is setting the `Request` timeout. 

Lastly, to be able to actually extend the channel timeout we'd need to always create a new channel, which is not desired. So a solution would be to set it very high by default, and make sure we always set the `Request` timeout to a default max if it was not specified before

cc @timvisee

**Comment by timvisee** [comment]
After some investigation from my end I came to the same conclusion.

I agree that the right approach is to remove (or make very high) the channel timeout and set it per request everywhere.

Thanks for sharing your findings!

**Comment by timvisee** [comment]
<https://github.com/qdrant/qdrant/pull/2771> implements the [above](https://github.com/qdrant/qdrant/pull/2748#issuecomment-1749541131) suggestion. When we merge it, I think we can merge this as well.

**Comment by coszio** [comment]
Now that #2771 has been merged, I've rebased this one, and manually tested the case of setting timeout to more than the configured one. It now works in both cases.

**Pull request merged**